module/Makefile

```diff
@@ -48,10 +48,17 @@ clean:
 	rm -rf *.o *.ko *.symvers *.mod.c .*.cmd Module.markers modules.order *.o.*
 	rm -rf .tmp_versions .rapiddisk.o.d *.unsigned *.sdtinfo.c .ctf/ .cache.mk
 
-.PHONY: dkms
-dkms:
-	sudo mkdir -pv /usr/src/rapiddisk-$(VERSION)/
-	sudo cp -v * /usr/src/rapiddisk-$(VERSION)/
-	sudo dkms add -m rapiddisk -v $(VERSION)
-	sudo dkms build -m rapiddisk -v $(VERSION)
-	sudo dkms install -m rapiddisk -v $(VERSION)
+dkms-install:
+ifeq ($(shell dkms status rapiddisk/$(VERSION) -k $(KVER) | grep '$(KVER)' >/dev/null && echo 0 || echo 1 ),0)
```

Comment:
I am testing this now (testing on CentOS7). Looks like we will still need to cp the source over to /usr/src/rapiddisk-$(VERSION):
```
dkms install rapiddisk/6.0 -k 3.10.0-957.21.3.el7.x86_64 .
Error! Could not find module source directory.
Directory: /usr/src/rapiddisk-6.0 does not exist.
make[1]: *** [dkms-install] Error 2
make[1]: Leaving directory `/root/rapiddisk/module'
```

Meaning, maybe we need to reinsert:

```
	sudo mkdir -pv /usr/src/rapiddisk-$(VERSION)/
	sudo cp -v * /usr/src/rapiddisk-$(VERSION)/
```

Unless I am overlooking something. Thoughts?

Reply:
I made some test. On CentOS7 I installed by hand the latest release of `dkms`, 2.7.1. On Ubuntu the provided version is 2.2.1.0. The documentation is a bit different between the versions, but long story short, there is no way I could find to use the same `dkms` syntax in both the environments and avoid the manual mkdir/copy.

So I believe your solution is the best one: first mkdir/copy, and then use dkms. After some hours this is the "best" `dkms` syntax I found:

    mkdir -pv /usr/src/rapiddisk-$(VERSION)/
    cp -v * /usr/src/rapiddisk-$(VERSION)/
    dkms install rapiddisk/$(VERSION) -k $(KVER) .

This way the module is compiled and installed, and we can still specify module version and kernel version. On uninstall we should not `rm -rf` cause the module might be in use on kernel versions other than the current.

Anyway `dkms` is a strange beast, on CentOS7 this syntax works: `dkms` copies the files and does all the magic:

    [shub@new-host module]$ sudo dkms install .
    
    Creating symlink /var/lib/dkms/rapiddisk/6.0/source ->
                     /usr/src/rapiddisk-6.0
    
    DKMS: add completed.
    
    Kernel preparation unnecessary for this kernel.  Skipping...
    
    Building module:
    cleaning build area...
    make -j4 KERNELRELEASE=3.10.0-957.21.3.el7.x86_64 -C /lib/modules/3.10.0-957.21.3.el7.x86_64/build M=/var/lib/dkms/rapiddisk/6.0/build...
    cleaning build area...
    
    DKMS: build completed.
    
    rapiddisk.ko.xz:
    Running module version sanity check.
    modinfo: ERROR: missing module or filename.
     - Original module
       - No original module exists within this kernel
     - Installation
       - Installing to /lib/modules/3.10.0-957.21.3.el7.x86_64/extra/
    
    rapiddisk-cache.ko.xz:
    Running module version sanity check.
    modinfo: ERROR: missing module or filename.
     - Original module
       - No original module exists within this kernel
     - Installation
       - Installing to /lib/modules/3.10.0-957.21.3.el7.x86_64/extra/
    Adding any weak-modules
    
    depmod...
    
    Backing up initramfs-3.10.0-957.21.3.el7.x86_64.img to /boot/initramfs-3.10.0-957.21.3.el7.x86_64.img.old-dkms
    Making new initramfs-3.10.0-957.21.3.el7.x86_64.img
    (If next boot fails, revert to initramfs-3.10.0-957.21.3.el7.x86_64.img.old-dkms image)
    dracut.........
    
    DKMS: install completed.

The drawback is that it is not possibile to specify kernel version or anything else: it stops working, as you noticed...

The `modinfo: ERROR: missing module or filename.` message is something I can't understand.

BTW, in `dkms.conf` you set 

    AUTOINSTALL="yes"

this means that if the dkms service is running, it will try to auto-install the module on every kernel you boot with. I was wondering if this was an intended behaviour. Quote:

    AUTOINSTALL=
        If this directive is set to yes then the service /etc/rc.d/init.d/dkms_autoinstaller will automatically  try  to
        install this module on any kernel you boot into.  See the section on dkms_autoinstaller for more information.

Regards

Reply:
autoinstall being enabled was the intended behavior.

> On uninstall we should not rm -rf cause the module might be in use on kernel versions other than the current.

Agreed.

I think the rest of it looks and works well. So long as we add the mkdir/cp to cover all implementations of dkms on the install, then we are in business. If you push another commit, I will do another sanity test and if all is good, I will merge.